providers/keycloak.go

```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
 	p.Group = group
 }
 
-func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	json, err := requests.New(p.ValidateURL.String()).
+func (p *KeycloakProvider) SetRoles(roles []string) {
+	p.Roles = roles
+}
+
+type keycloakUserInfo struct {
+	Email  string   `json:"email"`
+	Groups []string `json:"groups"`
+}
+
+func (p *KeycloakProvider) getUserInfo(ctx context.Context, s *sessions.SessionState) (*keycloakUserInfo, error) {
+	var userInfo keycloakUserInfo
+	err := requests.New(p.ValidateURL.String()).
 		WithContext(ctx).
 		SetHeader("Authorization", "Bearer "+s.AccessToken).
 		Do().
-		UnmarshalJSON()
+		UnmarshalInto(&userInfo)
 	if err != nil {
 		logger.Errorf("failed making request %s", err)
-		return "", err
+		return nil, fmt.Errorf("error getting user info: %v", err)
+	}
+	return &userInfo, nil
+}
+
+func getClientRoles(resourceAccess map[string]interface{}) ([]string, error) {
+	var clientRoles []string
+	for name, list := range resourceAccess {
+		scopes, ok := list.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("error parsing client roles from claims")
+		}
+		if roles, found := scopes["roles"]; found {
+			for _, r := range roles.([]interface{}) {
```

Comment:
Do you need to handle when this isn't a `[]interface{}` or have fallback code that somehow marshals non-slices if you get a singleton? We have to handle that case in the OIDC side: https://github.com/oauth2-proxy/oauth2-proxy/pull/936/commits/48baecb5c996d378f6871fa41a1c4824354dd083

Reply:
From what I've seen it always returns a slice even when if it only contains a single element, is there any other case that I'm missing?

Reply:
Cool - that makes sense that it is uniform since it is all provided by the singular Keycloak. OIDC is all over the place cause identity providers do some crazy stuff with `groups` since it isn't a totally official claim.